src/lsp-client.ts

```diff
@@ -18,7 +26,47 @@ export interface LspClient {
 }
 
 export class LspClientImpl implements LspClient {
-    constructor(protected connection: lsp.Connection) {
+    private clientCapabilities?: lsp.ClientCapabilities;
+
+    constructor(protected connection: lsp.Connection) {}
+
+    setClientCapabilites(capabilites: lsp.ClientCapabilities): void {
+        this.clientCapabilities = capabilites;
+    }
+
+    createProgressReporter(): ProgressReporter {
+        let workDoneProgress: Promise<lsp.WorkDoneProgressServerReporter> | undefined;
+        return {
+            begin: (message = '') => {
+                if (this.clientCapabilities?.window?.workDoneProgress) {
+                    workDoneProgress = this.connection.window.createWorkDoneProgress();
+                    workDoneProgress
+                        .then((progress) => {
+                            progress.begin(message);
+                        })
+                        .catch(() => {});
+                }
+            },
+            report: (message: string) => {
```

Comment:
Just an observation.
this report is not currently used in the project?

Because the ts service only has `EventTypes.ProjectLoadingStart`, `EventTypes.ProjectLoadingFinish` events. There is no event that can be used for this `report: (message: string)` function. Is that right?

Reply:
Yeah but I'll just leave it to make the interface future-proof since it's so little extra code.